File: decl-ts/tests/internal/engine_test.ts
// engine (tests/internal/checks.json): the engine's boundary through the
// single-module pipeline — quantities, references, $referrers, a cycle.
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { qevaluateUniverse } from '../../src/qengine/qeval.ts';
import { Env } from '../../src/semantics.ts';
import { initParser } from '../../src/node.ts';
import { parseSource } from '../../src/parse.ts';
import { isQ } from '../../src/semantics.ts';
import { runPipeline } from '../../src/pipeline.ts';
import { check, total, root } from '../common/check.ts';

await initParser();
{
  const q = runPipeline(
    parseSource('dimension Speed = Length / Time\nunit mps: Speed\noutput v: quantity<Speed> = 3km / 2s\n').decls,
  );
  const v = q.eng.resolveSegs(['v']);
  const r = runPipeline(
    parseSource(
      'type S = { name: string, inbound = $referrers(L, "target") }\ntype L = { source: ref<S>, target: ref<S> }\ntype Top = { services: S[], links: L[] }\nexport output top: Top = { services: [{ name: "a" }, { name: "b" }], links: [{ source: services[0], target: services[1] }] }\n',
    ).decls,
  );
  const ser = r.eng.serialize(r.env.roots.get('top'), 'top');
  check(
    'values',
    q.diags.length === 0 &&
      isQ(v) &&
      v.dim === 'Length*Time^-1' &&
      v.value === 1500 &&
      r.diags.length === 0 &&
      ser.includes('"source":"$.services[0]"') &&
      ser.includes('"inbound":["$.links[0]"]'),
    JSON.stringify({ v, ser, qd: q.diags, rd: r.diags }),
  );
}
{
  const p = runPipeline(parseSource('type T = { a = b, b = a }\nexport output t: T = {}\n').decls);
  check('cycle', p.diags.some((d) => d.code === 'E5007'), JSON.stringify(p.diags));
}
{
  const cases: { file: string; reuse: boolean }[] = JSON.parse(
    readFileSync(join(root, 'tests/internal/rounds.json'), 'utf8'),
  );
  for (const row of cases) {
    const decls = parseSource(readFileSync(join(root, row.file), 'utf8')).decls;
    const ref = runPipeline(decls);
    const env = new Env();
    env.load(decls);
    const { report, eng } = qevaluateUniverse([{ env }], { env });
    const ok = !ref.diags.some((d) => d.severity === 'error');
    const outputs = ok
      ? ref.env.outputs.filter((o) => ref.env.roots.has(o.name)).map((o) => ({
          name: o.name,
          json: ref.eng.serialize(ref.env.roots.get(o.name), o.name),
        }))
      : [];
    check(
      'reference_round_reuse',
      report.ok === ok &&
        JSON.stringify(report.outputs) === JSON.stringify(outputs) &&
        JSON.stringify(report.diagnostics) === JSON.stringify(ref.diags) &&
        (!row.reuse ||
          (eng.roundCache!.reusedRounds >= 3 && eng.roundCache!.retainedRecords >= 20)),
      row.file,
    );
  }
}
total();

File: decl-rs/tests/internal/engine_test.rs
//! engine: the engine's boundary through the single-module pipeline —
//! quantities, references, $referrers, a cycle.
use super::common::{get, read_json, text};
use decl_lang::parse::parse_source;
use decl_lang::pipeline::run_pipeline;
use decl_lang::qengine::qeval::qevaluate_universe;
use decl_lang::semantics::{Env, Seg, Value};

#[test]
fn values() {
    let q = run_pipeline(&parse_source("dimension Speed = Length / Time\nunit mps: Speed\noutput v: quantity<Speed> = 3km / 2s\n").decls);
    assert!(q.diags.is_empty(), "{:?}", q.diags);
    let v = q
        .eng
        .resolve_segs(&[Seg::Name("v".into())])
        .ok()
        .expect("v");
    assert!(
        matches!(&v, Value::Q { dim, value } if dim == "Length*Time^-1" && *value == 1500.0),
        "{v:?}"
    );
    let r = run_pipeline(&parse_source("type S = { name: string, inbound = $referrers(L, \"target\") }\ntype L = { source: ref<S>, target: ref<S> }\ntype Top = { services: S[], links: L[] }\nexport output top: Top = { services: [{ name: \"a\" }, { name: \"b\" }], links: [{ source: services[0], target: services[1] }] }\n").decls);
    assert!(r.diags.is_empty(), "{:?}", r.diags);
    let ser = r
        .eng
        .serialize(&r.env.root("top").expect("top"), "top", false);
    assert!(ser.contains("\"source\":\"$.services[0]\""), "{ser}");
    assert!(ser.contains("\"inbound\":[\"$.links[0]\"]"), "{ser}");
}

#[test]
fn cycle() {
    let p =
        run_pipeline(&parse_source("type T = { a = b, b = a }\nexport output t: T = {}\n").decls);
    assert!(
        p.diags.iter().any(|d| d.code.as_deref() == Some("E5007")),
        "{:?}",
        p.diags
    );
}

#[test]
fn reference_round_reuse() {
    let root = std::path::Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .unwrap();
    let Ok(Value::JArr(cases)) =
        read_json(&std::fs::read_to_string(root.join("tests/internal/rounds.json")).unwrap())
    else {
        panic!("rounds.json is a list");
    };
    for row in cases.iter() {
        let file = text(get(row, "file").unwrap());
        let parsed = parse_source(&std::fs::read_to_string(root.join(file)).unwrap());
        let reference = run_pipeline(&parsed.decls);
        let env = Env::new();
        env.load(&parsed.decls);
        let (report, eng) = qevaluate_universe(std::slice::from_ref(&env), &env, &[])
            .unwrap_or_else(|e| panic!("{}", e.0));
        let ok = !reference.diags.iter().any(|d| d.severity == "error");
        let outputs: Vec<_> = if ok {
            reference
                .env
                .outputs
                .borrow()
                .iter()
                .filter_map(|(name, _, _)| {
                    reference
                        .env
                        .root(name)
                        .map(|v| (name.clone(), reference.eng.serialize(&v, name, false)))
                })
                .collect()
        } else {
            vec![]
        };
        assert_eq!(report.ok, ok, "{file}");
        assert_eq!(report.outputs, outputs, "{file}");
        assert_eq!(
            report
                .diagnostics
                .iter()
                .map(|d| d.to_json(None))
                .collect::<Vec<_>>(),
            reference
                .diags
                .iter()
                .map(|d| d.to_json(None))
                .collect::<Vec<_>>(),
            "{file}"
        );
        if matches!(get(row, "reuse"), Some(Value::Bool(true))) {
            let cache = eng.round_cache.borrow();
            let cache = cache.as_ref().unwrap();
            assert!(cache.reused_rounds.get() >= 3, "{file}");
            assert!(cache.retained_records.get() >= 20, "{file}");
        }
    }
}
